### ISSUE TITLE
Upper version constraint in pubspec.yaml for flutter environment variable not working

### ISSUE THREAD

**Issue by neiljaywarner** [flutter/flutter#95472]
steps to reproduce
1. use current stable version flutter 2.8.1
2. in pubspec.yaml use   flutter: ">=2.8.0 <2.8.1" or  "2.8.0"

Expected behaviour - flutter pub get error 
---
```
The current Flutter SDK version is 2.8.1.

Because app_name requires Flutter SDK version <2.8.1, version solving failed.
```
Actual behaviour: pub get succeeds; vesrion constraint fails, overzealous dev or build server just ugpraded to a newer version of flutter without discussion or any testing at all or possibly even an easy OOTB way to see what version was built with

**Comment by darshankawar** [comment]
@neiljaywarner 
Is there any reason why you need to change it to `>=2.8.0 <2.8.1" or "2.8.0"` ?
On latest stable 2.8.1, the dart version supported is 2.15.1 (from `flutter doctor`) .

**Comment by neiljaywarner** [comment]
@darshankawar sorry I didn't see your question.
I want to pin it so if I go away for 6 to 18 months and come back I know what flutter version this project expects and will compile on.
Yes I know updating is important but so is keeping everyone on a project on the same tested version

**Comment by neiljaywarner** [comment]
Specifically I would like it to behave as expected consistently in both
directions and enforce upper constraint as well as lower constraint.

Not just force/remind devs to upgrade when the team decides to but disallow
them to race ahead.
For example ci should not be able to make a build with a flutter version
never tested by qa

....sent from my phone

**Comment by darshankawar** [comment]
The current version with upper constraint I have is `sdk: ">=2.8.1 <3.0.0"` which works well, indicating the min lower and max upper version constraint, so I am not entirely sure what your request here is.

**Comment by neiljaywarner** [comment]
What you have should disallow 3.0
0 it fails the constraint of bring less than or equal to 3.0
In fact 3.0 would pass and be compile and build except for the fact that
3.0 is a major version and would have breaking changes.

Honestly I want to be able to constrain to 2.8.1 for example so 2.8.0 fails
and 2.8.2 fails.

I don't ever want the project upgraded silently by random dev or random
server with no pr review or discussion.

This concept weeks with the lower constraint but upper constraint is broken
and as far as I could tell does nothing.

Thanks

....sent from my phone

**Comment by darshankawar** [comment]
Keeping it open and labeling for further input.

**Comment by neiljaywarner** [comment]
@darshankawar  is it clear my goal now?  I want to pin a specific version so neither ci/cd nor a new dev "accidentally" upgrades the flutter version for any builds they distribute to anybody and it won't compile at all on an incorrect flutter version. + @goderbauer

**Comment by maheshmnj** [comment]
cc: @christopherfujino

**Comment by christopherfujino** [comment]
Per https://dart.dev/tools/pub/pubspec#sdk-constraints, the way to do this is via the `environment` key, like:

```
environment:
  # This is the dart constraint
  sdk: '>=2.10.0 <3.0.0'
  # This is the Flutter SDK
  flutter: 2.8.0
```

From the link:

> A Flutter SDK constraint is satisfied only if pub is running in the context of the flutter executable, and the Flutter SDK’s version file matches the given version constraint. Otherwise, the package will not be selected.

However, this doesn't seem to be enforced when I locally try this.

**Comment by christopherfujino** [comment]
> @christopherfujino @darshankawar @maheshmnj yes, that's the whole point, it is supposed to work but doesn't.

@neiljaywarner you don't need to get snippy with me, I'm just documenting more context, to help solve this.

**Comment by neiljaywarner** [comment]
I am so sorry.
I did not mean to be snippy I was just concerned about clarity and I think
we are in agreement.
I love flutter and dart and I appreciate the response and I'm very sorry
please forgive me .



....sent from my phone

**Comment by christopherfujino** [comment]
I am renaming this issue not as a new feature proposal, but rather a bug fix to the existing pubspec.yaml flutter -> sdk constraint not working.

**Comment by christopherfujino** [comment]
Another filing of this in https://github.com/flutter/flutter/issues/80610

**Comment by christopherfujino** [comment]
Ok, in playing around with this further, it seems that lower bounds are respected (e.g. `flutter: '>=5.0.0'` **does** throw an error) but upper-bounds are not (e.g. `flutter: '<2.2.0'` does **not** throw an error). Also, exact pins are ignored (e.g. `flutter: '2.3.4'` does **not** throw).

I believe there is a historical reason for this. Also, I imagine it would be difficult to enable upper bounds, as this would be a breaking change. However, it seems exact pins would be very useful, and would be relatively safe to enable.

@Hixie @mit-mit do you remember the context around this?

**Comment by mit-mit** [comment]
Yeah, I remember. We changed this to enable the roll-out of Flutter 2; prior to that we had both a lower and upper bound, but when we went to Flutter 2 -- which wasn't a major breaking release from a semver perspective -- we would have "broken the world", so we removed the upper-constraint.

Pinning to a single SDK is a request I've heard fairly often. It can be useful for teams that want to ensure all team members are on the same SDK. Unfortunately a constraint like `flutter: x.y.z` means use x.y.z as the lower and upper constraint, and sine the upper constraint is ignored in Flutter it's only a lower constraint.

**Comment by christopherfujino** [comment]
> Yeah, I remember. We changed this to enable the roll-out of Flutter 2; prior to that we had both a lower and upper bound, but when we went to Flutter 2 -- which wasn't a major breaking release from a semver perspective -- we would have "broken the world", so we removed the upper-constraint.
> 
> Pinning to a single SDK is a request I've heard fairly often. It can be useful for teams that want to ensure all team members are on the same SDK. Unfortunately a constraint like `flutter: x.y.z` means use x.y.z as the lower and upper constraint, and sine the upper constraint is ignored in Flutter it's only a lower constraint.

Is the blocker to enabling `flutter: x.y.z` to exactly constrain the version a technical or policy one?

**Comment by neiljaywarner** [comment]
i also notice
  flutter: "2.10.5" # pubspec.yaml

becomes
  flutter: ">=2.10.5" # pubspec.lock

**Comment by simplifit-michael** [comment]
It is also a technical issue.
in my company we work on several project with several flutter versions.

eg:
one project is using flutter 3.3.2
the newest one is using flutter 3.7.0

=> the flutter_localizations of flutter 3.7.0 had breaking changes adding parameters. so when I want to use a string from the localisation it changed from being `S.of(context)!.stringwithparameter.replaceAll('${parameter}',value)` to `S.of(context)!.stringwithparameter(value)`
Also other dependencies have breaking changes if the flutter version is newer then the project defines.

Therefore a dev should be infomed when he uses a flutter version which is "too new"

hope that helps :)

**Comment by AlexV525** [comment]
Using a fixed Flutter version would be enough for cases that force CI/CD not to accidentally upgrade to a newer version of Flutter (mostly happened during a new stable version released but the project has not been migrated). However, enabling version constraints like `3.10.x` might be a little bit more helpful in a real-world project.

**Comment by guneyozsan** [comment]
We migrated to flutter recently and flutter version installed among team was one of the first team issues we hit. Setting up a new machine, or coordinating version upgrades among team would be difficult. I can also foresee debugging weirdness in CI if it silently gets a breaking version of flutter.

In our case, we don't essentially need a range but locking to a wildcard version could be nice (e.g. 3.19.x or 3.x.x).

Also, the range silently failing is hard to notice unless you notice the difference in the diff:
```
pubspec.yaml:
flutter: '>=3.19.0 <4.0.0'
pubspec.lock:
flutter: ">=3.19.0"
```

**Comment by cupidchan** [comment]
Hello, this issue was first opened in 2021. Is there any update so that we can fix the SDK and Flutter version?

**Comment by SharbelOkzan** [comment]
Facing the same issue. 
`pub get` returns no error when run with a different flutter version than the one set in pubspec.yaml  (I'm not using a range).

pubspec.yaml
```
environment:
  sdk: '>=3.0.5 <4.0.0'
  flutter: '3.22.1'
``` 
While using Flutter 3.22.3 still returns 0 exit code

**Comment by lbonsey1** [comment]
We are experiencing the same problem. Is this going to be addressed by the Flutter team soon?

**Comment by keyfayaz** [comment]
+1

**Comment by eseidel** [comment]
Hmm.  We were exploring respecting environment.flutter constraint in `shorebird` just now and then realized pub does not respect an exact revision of environment.flutter (but rather turns it into >=x.y.z rather than just x.yz). Could someone give me more context as to why it doesn't?  It seems this is the tracking issue for such (and I've now read all the comments) but I don't quite understand why it's this way.  If I sent a patch to pub to respect environment.flutter exact revisions could it be accepted?

**Comment by eseidel** [comment]
It appears the ignoring comes from https://github.com/dart-lang/pub/blob/99e84bc4c5e5fbdd683e4613f19eca31d82685e7/lib/src/pubspec.dart#L819

It is true that Flutter does not use semantic versioning for breaking releases.  But presumably we could still support upper bounds on environment.flutter. :)

**Comment by dsyrstad** [comment]
> It is true that Flutter does not use semantic versioning for breaking releases. But presumably we could still support upper bounds on environment.flutter. :)

...Or perhaps a flag in pubspec.yaml saying you want to enforce strict version checking for the flutter sdk. This would be backwards compatible with existing pubspecs. Along the lines of:

```yaml
environment:
  sdk: '>=3.5.0 <4.0.0'
  flutter: 3.29.2
  strict_flutter_version_checking: true
```

**Comment by eseidel** [comment]
It looks like this behavior was originally added in https://github.com/dart-lang/pub/commit/d74520ac58d072b007ae4e0bdf67639d97429166 Which appears to have been a request from Tim Sneath. 🤣  https://github.com/flutter/flutter/issues/68143

**Comment by eseidel** [comment]
My guess is what's happening here is the divergence of desires between packages and applications.  I expect that Tim's original request was so that not every flutter-using package in the Dart ecosystem would need an update just because Flutter announced Flutter 2, since by default ^1.0.0 will disallow usage of the package with Flutter 2 and thus the day that Flutter 2 was announced, you couldn't use it because all of your packages (which probably would work fine) suddenly would refuse to work with your version of Flutter because it was now not aligned with the ^1.0.0 constraint.

However this also meant that then Apps were no longer able to restrict the version of Flutter used, because (at least the way it was implemented) apps no longer had a way to specify they wanted exactly 1.0.0.

**Comment by eseidel** [comment]
Anyway, I think I'm stuck at this point.  I can't get a working test environment for pub locally (7 tests fail for me) and I have other demands on my time.  So I may look at this again (or someone else should feel free to attempt to 1. get pub tests working locally and 2. finish the posted patch).  My guess is that the proper solution here is going to be to one of:
* splitting behavior of "apps" vs. "packages" (not sure how that would work?)
* adding a new key to pubspec.yaml to be a real flutter constraint
* adding some new key to opt in to this behavior
* give up on this behavior and just make it so that when Flutter 4 is announced, all pub packages depending on flutter won't work with Flutter 4, because their constraints will be >=3 <4 🤷

**Comment by eseidel** [comment]
To restate (for clarity) the problem that I understand this bug to be about.

*There is no official way to record "this app should be built with version x.y.z of Flutter".*

As currently implemented, if one use the environment.flutter constraint, one would, at minimum, need to look at the pubspec.yaml rather than pubspec.lock because the spec might say 1.2.3 but the lock will say >=1.2.3.

This arrises because of the workaround added in https://github.com/dart-lang/pub/commit/d74520ac58d072b007ae4e0bdf67639d97429166.  That workaround was targeted a *libraries* to avoid a library which was OK with Fluter 1.x suddenly needing to re-publish when Flutter 2.x was released, even though 2.x had no breaking changes for that library.  When the flutter/dart ecosystem was smaller, that was a very bad thing for Flutter 2 release day. :)

There are lots of 3p attempts at adding ways to express "this app should be built with flutter x.y.z". fvm, puro, etc. have per-package "lock" files for a flutter version. There is just no "flutter official" place to put this information.

How we (shorebird) got involved here was simply that we were attempting to teach `shorebird` to determine the version of Flutter to use for building a project (as part of fixing https://github.com/shorebirdtech/shorebird/issues/2529).  When we realized the environment.flutter method was unreliable, we had 3 options: we (shorebird) make yet another way, fix this one, or bless one of these 3p ways as being the "standard".

It's not a huge blocker for us. Just attempting to re-summarize this long issue and explain why my attention got drawn to this (briefly) this week.

**Comment by neiljaywarner** [comment]
Oh two other quick comments 
1) I'm working with a very large Enterprise right now and concerns about upgrading flutter have come up 
2) It's a really big problem in my opinion that since you can't pin the flutter version if you set aside something and come back to at 6 or 12 months later, you don't even know what flutter version it was built with

**Comment by AlexV525** [comment]
Some unthoughtful idea: How about introducing "!x.y.z" for specific version constraint pin?

**Comment by sigurdm** [comment]
After discussing this offline we ended up agreeing that we will change pub to respect the upper bound flutter constraint for any root pubspec (that is, the pubspecs of the workspace).

That is a breaking change, but most likely not very disruptive, as usually you can change the pubspecs of the workspace, while it is harder to.

**Comment by eseidel** [comment]
Sounds good. I think the only time it will be "disruptive" is when Google launches Flutter 4.0 and it takes months for ones favorite packages to work with 4.0.  (As I recall that was the original motivation for this current behavior was launching Flutter 2.0.)

**Comment by mit-mit** [comment]
Right, but that is a lot less disruptive for a root pubspec (for those the developer likely owns the source code and can quickly update it), than it is for the many packages that are listed as dependencies in that root pubspec. These packages often come from pub, and would need a coordinated effort across the ecosystem to get the contraints updated.

**Comment by sigurdm** [comment]
Fixed by https://github.com/dart-lang/pub/pull/4595
